### PR TITLE
Write environment variables to unique file for msvc toolchain

### PR DESCRIPTION
**Pull request by niranjanyardi**
b/297227714

Make a separate environment file per instance of msvc toolchain template.
This fixes a deadlock where multiple toolchains using the same msvc_toolchain template try to write to the same environment file.

### CODE REVIEW

**Review by andrewsavage1 — APPROVED** [review]

File: build/toolchain/win/msvc_toolchain.gni [hunk 1]
```diff
@@ -16,6 +16,7 @@ import("//build/toolchain/cc_wrapper.gni")
 import("//build/toolchain/toolchain.gni")
 
 tool_wrapper_path = rebase_path("tool_wrapper.py", root_build_dir)
+assert(use_cobalt_customizations)
```
Comment (andrewsavage1) [inline build/toolchain/win/msvc_toolchain.gni:19]: Add a helpful message here and let's run it immediately (before even imports occur)

File: build/toolchain/win/msvc_toolchain.gni [hunk 2]
```diff
@@ -51,7 +51,12 @@ template("msvc_toolchain") {
 
   nul = "$0x00"
   env_block = string_join(nul, environment_key_value_pairs) + nul
-  write_file("$root_build_dir/environment.$target_cpu", env_block)
+  if (!use_cobalt_customizations) {
+    write_file("$root_build_dir/environment.$target_cpu", env_block)
+  } else {
+     write_file("$root_build_dir/$target_name/environment.$target_cpu",
```
[the remark this answers is not in the document]
Comment (niranjanyardi) [reply build/toolchain/win/msvc_toolchain.gni]: dunno which other GN variable we can leverage.

File: build/toolchain/win/msvc_toolchain.gni [hunk 3]
```diff
@@ -51,7 +52,11 @@ template("msvc_toolchain") {
 
   nul = "$0x00"
   env_block = string_join(nul, environment_key_value_pairs) + nul
-  write_file("$root_build_dir/environment.$target_cpu", env_block)
+
+  # Refer b/297227714, if default toolchain shares name with another toolchain, there's
```
Comment (andrewsavage1) [inline build/toolchain/win/msvc_toolchain.gni:56]: Update this comment slightly—adding $target_name does not fix the issue when two toolchains share the same name, that's when the deadlock could occur. This is supposed to be a note for future devs to watch out for a case like this and hopefully make a hard-to-debug problem a little easier

Comment (niranjanyardi) [reply build/toolchain/win/msvc_toolchain.gni:56]: updated.

File: build/toolchain/win/msvc_toolchain.gni (hunk 1 above)
Comment (niranjanyardi) [reply build/toolchain/win/msvc_toolchain.gni]: added before imports. 
I think it's best to let the developer see the definition use_cobalt_customizations : as that can change over time.

File: build/toolchain/win/msvc_toolchain.gni [hunk 4]
```diff
@@ -12,11 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+assert(use_cobalt_customizations)
```
Comment (andrewsavage1) [inline build/toolchain/win/msvc_toolchain.gni:15]: net: add a newline after this

File: build/toolchain/win/msvc_toolchain.gni [hunk 5]
```diff
@@ -51,7 +53,10 @@ template("msvc_toolchain") {
 
   nul = "$0x00"
   env_block = string_join(nul, environment_key_value_pairs) + nul
-  write_file("$root_build_dir/environment.$target_cpu", env_block)
+
+  # Refer b/297227714, if default toolchain shares name with another toolchain, there's
```
Comment (andrewsavage1) [inline build/toolchain/win/msvc_toolchain.gni:57]: Replace with: "If the default toolchain shares its name with another toolchain, there could be a deadlock where both toolchains try to write to the same environment file. See b/297227714 for more context."

**Review by andrewsavage1 — APPROVED** [review]
Please manually run `gn format` on this file (it won't get run in pre-commit checks as it's under //build)

File: build/toolchain/win/msvc_toolchain.gni (hunk 4 above)
Comment (niranjanyardi) [reply build/toolchain/win/msvc_toolchain.gni:15]: done

File: build/toolchain/win/msvc_toolchain.gni (hunk 5 above)
Comment (niranjanyardi) [reply build/toolchain/win/msvc_toolchain.gni]: done

File: build/toolchain/win/msvc_toolchain.gni (hunk 4 above)
Comment (niranjanyardi) [reply build/toolchain/win/msvc_toolchain.gni:15]: done

**Comment by niranjanyardi** [comment]
> Please manually run `gn format` on this file (it won't get run in pre-commit checks as it's under //build)

done.

Should we move out our files to a different folder ?

**Pull request merged**